submissions/src/main/kotlin/gov/cdc/prime/reportstream/submissions/controllers/SubmissionController.kt

```diff
@@ -0,0 +1,185 @@
+            if (headerValidationResult != null) {
+                logger.warn("Header validation failed: $headerValidationResult")
+                return headerValidationResult
+            }
+
+            // Convert data to ByteArray
+            val dataByteArray = data.toByteArray()
+            logger.debug("Converted report data to ByteArray")
+
+            // Upload to blob storage
+            val blobContainerClient = blobServiceClient.getBlobContainerClient(containerName)
+            val blobClient = blobContainerClient.getBlobClient(formBlobName(reportId, headers))
+            blobClient.upload(dataByteArray.inputStream(), dataByteArray.size.toLong())
+            logger.info("Uploaded report to blob storage: blobUrl=${blobClient.blobUrl}")
+
+            // Create the message for the queue
+            val message = mapOf(
+                "reportId" to reportId.toString(),
+                "blobUrl" to blobClient.blobUrl,
+                "headers" to filterHeaders(headers)
+            )
+            val objectMapper = jacksonObjectMapper().registerModule(JavaTimeModule())
+            val messageString = objectMapper.writeValueAsString(message)
+            logger.debug("Created message for queue")
+
+            // Upload to Queue
+            val queueClient = queueServiceClient.getQueueClient(queueName)
+            queueClient.createIfNotExists()
+            queueClient.sendMessage(messageString)
+            logger.info("Sent message to queue: queueName=$queueName")
+
+            // Insert into Table
+            // TableEntity() sets PartitionKey and RowKey. Both are required by azure and combine to create the PK
+            val tableEntity = TableEntity(reportId.toString(), reportId.toString())
+            val tableProperties = mapOf(
+                "report_received_time" to reportReceivedTime.toString(),
+                "report_accepted_time" to reportReceivedTime.toString(), // Will be updated when the report is accepted
+                "report_id" to reportId.toString(),
+                "status" to status
+            )
+            tableClient.createEntity(tableEntity.setProperties(tableProperties))
+            logger.info("Inserted report into table storage: reportId=$reportId")
+
+            // Create and publish custom event
+            val reportReceivedEvent = ReportReceivedEvent(
+                timeStamp = reportReceivedTime,
+                reportId = reportId,
+                parentReportId = reportId,
+                rootReportId = reportId,
+                headers = filterHeaders(headers),
+                sender = clientId!!,
+                senderIP = headers["x-azure-clientip"].toString(),
+                fileSize = headers["content-length"].toString(),
+                blobUrl = blobClient.blobUrl
+            )
+            logger.debug("Created ReportReceivedEvent")
+
+            // Log to Application Insights
+            telemetryClient.trackEvent(
```

Comment:
Maybe as a follow up, but we should move the azure event service over to shared, since all the event stuff should work the same way